src/backend/distributed/README.md

```diff
@@ -1955,12 +1955,153 @@ The distributed transactionId and backend/PID mapping is done via BackendData st
+
+The "consistency" here primarily refers to two things:
+
+1. **Order of Operations on Replicated Tables**: In a replicated table setup, the same data exists on multiple nodes. The function aims to make sure that any updates, deletes, or inserts happen in the same order on all copies of the data (replicas). This way, all the replicas stay in sync.
+
+2. **Preventing Distributed Deadlocks**: When you're running multi-shard operations that update the data, you can run into distributed deadlocks if operations on different nodes lock shards in a different order. This function ensures that the locks are acquired in a specific, consistent order, thus minimizing the risk of deadlocks.
+
+There are also options (`GUCs`) to relax these locking mechanisms based on the user's needs, but they come with the trade-off of potentially reduced consistency or increased risk of deadlocks.
+
+So, in summary, this function is about acquiring the right kind of lock based on what the operation is doing and what kind of table it's affecting, all to ensure that the data stays consistent and that operations are executed safely.
+
+
+#### Citus Advisory Locks 2: Metadata locks
+
+The second class of advisory locks referred as `metadata locks` in the code. See `AcquireMetadataLocks()` and `LockShardDistributionMetadata()` functions.
+
+The main goal is to prevent concurrent placement changes and query executions. Essentially ensure that the query execution always works on the accurate placement metadata (e.g., shard placements).
+
+Citus always acquire `Metadata Lock`s for shard moves and shard splits, irrespetive of blocking vs non-blocking operations. For blocking operations, the locks are held from the start of the operation whereas for non-blocking operations the locks are held briefly at the end right before metadata is updated.
+
+Citus always acquires `Metadata Lock`s for modification queries, at the `CitusBeginModifyScan` such that it serializes query modification with placement metadata changes. A modification query would always see up-to date metadata for the placements involved. Otherwise, the modification might get lost.
+
+Citus does not acquire `Metadata Lock`s for `SELECT` queries. The main reason is that `SELECT`s are often long-running and would hold up the move. Instead, we allow `SELECT` commands to operate on the old placements in case of a concurrent shard move. The `SELECT` commands would already see the snapshot of the shard(s) when the `SELECT` started. So, there is no difference in terms of query correctness. In practice, the lack of locks caused deadlock & failures when the shard got dropped from underneath while `SELECT` is running. So we added deferred drop (see `Resource cleanup`), which leaves the shards around for a little longer, so the `SELECT` can calmly finish.
```

Comment:
structure is a bit confusing

> Instead, we allow `SELECT` commands to operate on the old placements in case of a concurrent shard move.

> In practice, the lack of locks caused deadlock & failures when the shard got dropped from underneath while `SELECT` is running.

it kind of sounds like allowing the SELECT commands to operate on the old placements caused deadlock & failures in practice.


thinking about this more: holding up the move was probably not the reason why we didn't add metadata locks, since the DROP TABLE at the end of the move would still be blocked if there was a long-running SELECT. I think one reason is that they query planner already assigns tasks to placement, and in case of prepared statements the query planner might run in a separate transaction. In general, there was not one specific point where those locks could be placed.

Perhaps we should not try to do the archeology and just say:

> Citus does not acquire `Metadata Lock`s for `SELECT` queries. The main reason is that `SELECT`s are often long-running and would hold up the move. Instead, we allow `SELECT` commands to operate on the old placements in case of a concurrent shard move. The `SELECT` commands would already see the snapshot of the shard(s) when the `SELECT` started. So, there is no difference in terms of query correctness. We then later drop the old placements via "deferred drop" (see Resource cleanup).